rgbd/src/client_shm.cpp

```diff
@@ -0,0 +1,134 @@
+#include "rgbd/client_shm.h"
+
+#include "rgbd/image.h"
+
+#include <ros/init.h>
+#include <ros/time.h>
+
+#include <sensor_msgs/CameraInfo.h>
+
+#include <boost/interprocess/sync/scoped_lock.hpp>
+
+namespace ipc = boost::interprocess;
+
+namespace rgbd
+{
+
+// ----------------------------------------------------------------------------------------------------
+
+ClientSHM::ClientSHM() : buffer_header_(nullptr), image_data_(nullptr)
+{
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+ClientSHM::~ClientSHM()
+{
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+bool ClientSHM::intialize(const std::string& server_name, float timeout)
+{
+    ros::Time start = ros::Time::now();
+    ros::Time now;
+    ros::Duration d(0.1);
+    do
+    {
+        try
+        {
+            std::string server_name_cp = server_name;
+            std::replace(server_name_cp.begin(), server_name_cp.end(), '/', '-');
+
+            // Open already created shared memory object.
+            shm_ = ipc::shared_memory_object(ipc::open_only, server_name_cp.c_str(), ipc::read_write);
+
+            mem_buffer_header_ = ipc::mapped_region(shm_, ipc::read_write, 0, sizeof(BufferHeader));
+            mem_image_ = ipc::mapped_region(shm_, ipc::read_only, sizeof(BufferHeader));
+
+            buffer_header_ = static_cast<BufferHeader*>(mem_buffer_header_.get_address());
+            image_data_ = static_cast<uchar*>(mem_image_.get_address());
+
+            rgb_data_size_ = static_cast<uint64_t>(buffer_header_->rgb_width * buffer_header_->rgb_height * 3);
+            depth_data_size_ = static_cast<uint64_t>(buffer_header_->depth_width * buffer_header_->depth_height * 4);
+
+            sequence_nr_ = 0;
+            return true;
+        }
+        catch(ipc::interprocess_exception &ex)
+        {
+//            std::cout << ex.what() << std::endl;
+        }
+        d.sleep();
+        now = ros::Time::now();
+     }
+     while (ros::ok() && (now - start).toSec() < static_cast<double>(timeout));
+
+    return false;
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+bool ClientSHM::initialized()
+{
+    return (buffer_header_ != nullptr);
+}
+
+// ----------------------------------------------------------------------------------------------------
+
+bool ClientSHM::nextImage(Image& image)
+{
+    if (!initialized())
+        return false;
+
+    ipc::scoped_lock<ipc::interprocess_mutex> lock(buffer_header_->mutex);
```

Comment:
Probably out-of-scope for this MR, but could this be implemented without a mutex (e.g., pointer swapping)?

Reply:
Pointer swapping?

Reply:
It would mean way too many changes for this MR.